src/bios.rs

```diff
@@ -1,627 +1,625 @@
+
+/// (`swi 0x00`) Performs a "soft reset" of the device.
+///
+/// Loads `r14` based on the `u8` value at `0x0300_7FFA`:
+/// * zero: `0x0800_0000` (ROM)
+/// * non-zero: `0x0200_0000` (EWRAM)
+///
+/// Then resets the following memory and registers:
+/// * `0x300_7E00` ..= `0x300_7FFF`: zeroed
+/// * `r0` ..= `r12`: zeroed
+/// * `sp_usr`: `0x300_7F00`
+/// * `sp_irq`: `0x300_7FA0`
+/// * `sp_svc`: `0x300_7FE0`
+/// * `lr_svc`, `lr_irq` : zeroed
+/// * `spsr_svc`, `spsr_irq`: zeroed
+///
+/// Then jumps to the `r14` address. This never returns.
+pub unsafe fn SoftReset() -> ! {
+  asm!("swi 0x00", options(noreturn))
 }
-newtype! {
-  /// Flags for use with `register_ram_reset`.
-  RegisterRAMResetFlags, u8
+
+/// (`swi 0x01`) Resets RAM and/or IO registers
+///
+/// * Note that if the IWRAM flag is used it doesn't reset the final `0x200`
+///   bytes of IWRAM. Instead, those bytes are reset during a call to the
+///   [`SoftReset`] function.
+/// * BIOS Bug: Data in `SIODATA32` is always destroyed, even if the `sio` flag
+///   is not set.
+#[inline]
+#[instruction_set(arm::t32)]
+pub unsafe fn RegisterRamReset(flags: crate::mmio_types::ResetFlags) {
+  asm!("swi 0x01",
+    inlateout("r0") flags.0 => _,
+    out("r1") _,
+    out("r3") _,
+    options(nomem, nostack)
+  )
 }
-#[allow(missing_docs)]
-impl RegisterRAMResetFlags {
-  phantom_fields! {
-    self.0: u8,
-    ewram: 0,
-    iwram: 1,
-    palram: 2,
-    vram: 3,
-    oam: 4,
-    sio: 5,
-    sound: 6,
-    other_io: 7,
-  }
+
+/// (`swi 0x02`) Halts the CPU until an interrupt request occurs.
+///
+/// The CPU is placed into low-power mode, while other parts (video, sound,
+/// timers, serial, keypad) continue to operate. This mode only terminates when
+/// one of the enabled interrupts is requested.
+///
+/// This halt state uses [`IE`] to determine what interrupts to allow, but the
+/// [`IME`] value is ignored (interrupts can occur even if `IME` is `false`).
+#[inline]
+#[instruction_set(arm::t32)]
+pub unsafe fn Halt() {
+  asm!("swi 0x02",
+    out("r0") _,
+    out("r1") _,
+    out("r3") _,
+    options(nomem, nostack)
+  )
 }
 
-/// (`swi 0x02`) Halts the CPU until an interrupt occurs.
-///
-/// Components _other than_ the CPU continue to function. Halt mode ends when
-/// any enabled interrupt triggers.
-#[inline(always)]
-pub fn halt() {
-  #[cfg(not(target_arch = "arm"))]
-  {
-    unimplemented!("This function is not supported on this target.")
-  }
-  #[cfg(target_arch = "arm")]
-  {
-    unsafe {
-      asm!("swi 0x02");
-    }
-  }
+/// (`swi 0x03`) Puts the CPU in a *very* low power state.
+///
+/// While stopped, the CPU, Sound, Video, SIO-shift-clock, DMA, and Timers are
+/// all disabled.
+///
+/// The system can return from this state only if there is an interrupt from the
+/// Keypad, Game Pak, or General-Purpose-SIO.
+///
+/// Before calling Stop you are advised to disable the Video to reduce battery
+/// usage, otherwise it just freezes.
+#[inline]
```

Comment:
The various BIOS calls are marked `inline`, as well as using THUMB code.

How does this interact with calls from ARM code?

Because of how the `swi` instruction works, and how the BIOS interprets the swi-number, BIOS calls from ARM-mode must shift the swi-number left by 16 bits.

Is this isn't considered, all BIOS calls from ARM-mode will behave like swi 0 (SoftReset).

If Rust in fact doesn't do cross-ISA inlining, this code is fine. Otherwise, these functions should likely be `inline(never)`, or a different solution should be found to ensure correct behaviour in ARM-mode.

Reply:
Rust doesn't do cross-isa inline because llvm itself doesn't do cross-isa inline.

also, each of these is marked as t32 code. the reason is that you can do a build that defaults to either a32 or t32, but of course these particular functions must be in t32.

so what happens in total is that a program built with t32 as the default code state will be able to inline all of these, and a program with a32 as default won't inline any of these.

presumably if we wanted to support a32 more we'd need a separate module like bios_a32.rs which contained a32 declarations of all these. however, strong a32 support is not currently a goal.

Reply:
I see. That looks good to me then